politeiad/backendv2/tstorebe/store/mysql/encrypt.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/decred/politeia/util"
+	"github.com/marcopeereboom/sbox"
+	"golang.org/x/crypto/argon2"
+)
+
+// salt creates a random salt and saves it to the kv store. Subsequent calls to
+// this function will return the existing salt.
+func (s *mysql) salt(size int) ([]byte, error) {
+	saltKey := "salt"
+
+	// Check if a salt already exists in the database
+	blobs, err := s.Get([]string{saltKey})
+	if err != nil {
+		return nil, fmt.Errorf("get: %v", err)
+	}
+	salt, ok := blobs[saltKey]
+	if ok {
+		// Salt already exists
+		log.Debugf("Salt found in kv store")
+		return salt, nil
+	}
+
+	// Salt doesn't exist yet. Create one and save it.
+	salt, err = util.Random(size)
+	if err != nil {
+		return nil, err
+	}
+	kv := map[string][]byte{
+		saltKey: salt,
+	}
+	err = s.Put(kv, false)
+	if err != nil {
+		return nil, fmt.Errorf("put: %v", err)
+	}
+
+	log.Debugf("Salt created and saved to kv store")
+
+	return salt, nil
+}
+
+// aragon2idKey derives a 32 byte aragon2id key from the provided password.
+// The salt is generated the first time the key is derived and saved to the kv
+// store. Subsequent calls to this fuction will use the existing salt.
+func (s *mysql) argon2idKey(password string) (*[32]byte, error) {
+	var (
+		pass           = []byte(password)
+		saltLen int    = 16 // In bytes
+		time    uint32 = 1
+		memory  uint32 = 64 * 1024 // 64 MB
+		threads uint8  = 4         // Number of available CPUs
+		keyLen  uint32 = 32        // In bytes
+	)
```

Comment:
The parameters should be save with the salt. Create a JSON struct that stores salt + params and add the encrypted blob to it. That way we can deal with upgrades and shit later.

Reply:
Done

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/store/mysql/encrypt.go#L20-L108

Reply:
Not sure what you mean by "add the encrypted blob to it".

Reply:
That code is a bit awkward to read. Can we create either a if exists {} else {} format that does everything inside their respective scope, or maybe better is to instantiate the structure and the fill it with values from the kv store or newlygenerated then let the code fallthrough to the common element.